lib/ClusterShell/Worker/Rsh.py

```diff
@@ -59,8 +60,25 @@ def _build_cmd(self):
         cmd_l.append("%s" % self.key)  # key is the node
         cmd_l.append("%s" % self.command)
 
+        # rsh does not properly return exit status
+        # force the exit status to be printed out
+        cmd_l.append("; echo XXRETCODE: $?")
+
         return (cmd_l, None)
 
+    def _on_nodeset_msgline(self, nodes, msg, sname):
+        """local wrapper over _on_node_msgline that can also handle nodeset"""
+        match = re.search("^XXRETCODE: (\d+)$", msg.decode("utf-8"))
+        if match:
+            self.rsh_rc = int(match.group(1))
+        self.worker._on_node_msgline(nodes, msg, sname)
```

Comment:
add a `else` above, unless you will call _on_node_msgline for output like` XXRETCODE:` and this will be visible by users.

Also, call the _on_node_msgline() from the parent class here:

    ExecClient._on_nodeset_msgline(self, nodes, msg, sname)

That's important because, as you can see in this function, there is a special logic there.

Reply:
Done!